File: src/my_py_pkg/my_py_pkg/hardware_status_publisher.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from my_robot_interfaces.msg import HardwareStatus


class HardwareStatusPublisherNode(Node):

    def __init__(self):
        super().__init__("hardware_status_publisher")

        self.hw_publisher_ = self.create_publisher(
            HardwareStatus,
            "hardware_status",
            10
        )

        self.timer_ = self.create_timer(
            1.0,
            self.publish_hardware_status
        )

    def publish_hardware_status(self):

        msg = HardwareStatus()

        msg.temperature = 63.33
        msg.are_motors_ready = True
        msg.debug_message = "Nothing to report"

        self.hw_publisher_.publish(msg)

        self.get_logger().info(
            f"Temperature: {msg.temperature} °C | "
            f"Motors Ready: {msg.are_motors_ready} | "
            f"Debug: {msg.debug_message}"
        )


def main(args=None):

    rclpy.init(args=args)

    node = HardwareStatusPublisherNode()

    rclpy.spin(node)

    node.destroy_node()

    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/hardware_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/hardware_status.hpp"

class HardwareStatusPublisher : public rclcpp::Node   
{
public:
  HardwareStatusPublisher() : Node("hardware_status_publisher")   
  {
    publisher_ = this->create_publisher<my_robot_interfaces::msg::HardwareStatus>("hardware_status", 10);   
    timer_ = this->create_wall_timer(
      std::chrono::seconds(1),
      std::bind(&HardwareStatusPublisher::publish_hardware_status, this));
    RCLCPP_INFO(this->get_logger(), "C++ ROS 2 node has started");
  }

private:
  void publish_hardware_status()
    {   
      auto message = my_robot_interfaces::msg::HardwareStatus();
      message.temperature = 42.0;
      message.are_motors_ready = true;
      message.debug_message = "All systems operational";
      publisher_->publish(message);
      RCLCPP_INFO(this->get_logger(), "Published hardware status: '%s'", message.debug_message.c_str());
    }

  rclcpp::Publisher<my_robot_interfaces::msg::HardwareStatus>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<HardwareStatusPublisher>());
  rclcpp::shutdown();
  return 0;
}

